hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/DatanodeManager.java

```diff
@@ -1755,12 +1755,24 @@ private BlockRecoveryCommand getBlockRecoveryCommand(String blockPoolId,
           LOG.info("Skipped stale nodes for recovery : "
               + (storages.length - recoveryLocations.size()));
         }
-        recoveryInfos = DatanodeStorageInfo.toDatanodeInfos(recoveryLocations);
       } else {
-        // If too many replicas are stale, then choose all replicas to
+        // If too many replicas are stale, then choose live replicas to
         // participate in block recovery.
-        recoveryInfos = DatanodeStorageInfo.toDatanodeInfos(storages);
+        recoveryLocations.clear();
+        storageIdx.clear();
+        for (int i = 0; i < storages.length; ++i) {
+          if (storages[i].getDatanodeDescriptor().isAlive()) {
+            recoveryLocations.add(storages[i]);
+            storageIdx.add(i);
+          }
+        }
+        assert recoveryLocations.size() > 0 : "recoveryLocations size should be > 0";
```

Comment:
Is this assert necessary here, or `recoveryLocations` size could be 0 if all DataNodes are not alive?

Reply:
Check the code again. when processing handleHeartbeat executes getBlockRecoveryCommand, the datanode  should be in the live state at this time, so the size of recoveryLocations is at least 1.
so here maybey remove this assert logic.